File: apps/ios/Voiid/Voiid/Networking/MuteStore.swift
//
//  MuteStore.swift
//  Voiid
//
//  Which conversations are muted, and until when.
//
//  ── WHY THE APP GROUP AND NOT UserDefaults.standard ─────────────────────────────
//  A muted conversation must stay silent when the app is not running — which is exactly
//  when it matters. Incoming messages are decrypted and presented by the NSE, a SEPARATE
//  PROCESS, and that process cannot see the app's private defaults. Storing mute state
//  anywhere else means it works only while the app is in the foreground, which is the one
//  case where the user can already see the message.
//
//  ── EXPIRY IS COMPUTED, NOT SCHEDULED ───────────────────────────────────────────
//  A mute has an end DATE; nothing has to fire when it passes. Scheduling an un-mute would
//  mean a timer that must survive a terminated app, a reboot, and a restore — all to
//  reproduce what a date comparison answers for free. `isMuted` reads the clock.
//

import Foundation

enum MuteStore {

    /// How long a mute lasts. `always` has no end date.
    enum Duration: String, CaseIterable, Identifiable {
        case eightHours, oneDay, oneWeek, always

        var id: String { rawValue }

        var title: String {
            switch self {
            case .eightHours: return "8 hours"
            case .oneDay:     return "24 hours"
            case .oneWeek:    return "1 week"
            case .always:     return "Until I turn it back on"
            }
        }

        var seconds: TimeInterval? {
            switch self {
            case .eightHours: return 8 * 3600
            case .oneDay:     return 24 * 3600
            case .oneWeek:    return 7 * 24 * 3600
            case .always:     return nil
            }
        }
    }

    private static let key = "voiid.muted.conversations"

    /// Shared with the NSE. Falls back to standard defaults only so a mis-provisioned build
    /// still behaves sanely in the foreground rather than crashing.
    private static var store: UserDefaults {
        UserDefaults(suiteName: AppGroup.identifier) ?? .standard
    }

    /// `conversationId -> end timestamp`. A value of 0 means "always".
    private static var table: [String: Double] {
        get { store.dictionary(forKey: key) as? [String: Double] ?? [:] }
        set { store.set(newValue, forKey: key) }
    }

    static func mute(_ conversationId: String, for duration: Duration) {
        var t = table
        t[conversationId] = duration.seconds.map { Date().timeIntervalSince1970 + $0 } ?? 0
        table = t
    }

    static func unmute(_ conversationId: String) {
        var t = table
        t.removeValue(forKey: conversationId)
        table = t
    }

    /// Is this conversation silent right now?
    ///
    /// Expired entries are cleaned up as they are read rather than swept: the read already
    /// has the answer, and a sweep would need somewhere to run from.
    static func isMuted(_ conversationId: String) -> Bool {
        guard let until = table[conversationId] else { return false }
        if until == 0 { return true }                      // always
        if Date().timeIntervalSince1970 < until { return true }
        unmute(conversationId)
        return false
    }

    /// When the mute ends, for the UI to say so. Nil when not muted or muted forever.
    static func mutedUntil(_ conversationId: String) -> Date? {
        guard let until = table[conversationId], until != 0 else { return nil }
        let date = Date(timeIntervalSince1970: until)
        return date > Date() ? date : nil
    }
}

File: apps/android/app/src/main/java/com/voiid/app/net/MuteStore.kt
package com.voiid.app.net

import android.content.Context

/**
 * Which conversations are muted, and until when. Twin of iOS `MuteStore.swift`.
 *
 * EXPIRY IS COMPUTED, NOT SCHEDULED: a mute stores an end time and [isMuted] reads the clock,
 * so nothing has to fire when it passes — it survives process death and reboots for free.
 * Stored in app-private prefs, which the FCM service (same process) reads before posting.
 */
object MuteStore {
    enum class Duration(val title: String, val millis: Long?) {
        EIGHT_HOURS("8 hours", 8 * 3_600_000L),
        ONE_DAY("24 hours", 24 * 3_600_000L),
        ONE_WEEK("1 week", 7 * 24 * 3_600_000L),
        ALWAYS("Until I turn it back on", null),
    }

    private const val PREFS = "voiid.muted.conversations"

    private fun prefs(ctx: Context) = ctx.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    /** 0 means "always". */
    fun mute(ctx: Context, conversationId: String, duration: Duration) {
        val until = duration.millis?.let { System.currentTimeMillis() + it } ?: 0L
        prefs(ctx).edit().putLong(conversationId, until).apply()
    }

    fun unmute(ctx: Context, conversationId: String) {
        prefs(ctx).edit().remove(conversationId).apply()
    }

    /** Expired entries are cleaned up as they are read. */
    fun isMuted(ctx: Context, conversationId: String): Boolean {
        val p = prefs(ctx)
        if (!p.contains(conversationId)) return false
        val until = p.getLong(conversationId, -1L)
        if (until == 0L) return true
        if (System.currentTimeMillis() < until) return true
        unmute(ctx, conversationId)
        return false
    }

    /** When the mute ends; null when not muted or muted forever. */
    fun mutedUntil(ctx: Context, conversationId: String): Long? {
        val until = prefs(ctx).getLong(conversationId, -1L)
        return until.takeIf { it > System.currentTimeMillis() }
    }

    fun clear(ctx: Context) = prefs(ctx).edit().clear().apply()
}
